### ISSUE TITLE
NVDA Remote with "Attempt to cancel speech for expired focus events" speech fails

### ISSUE THREAD

**Issue by cary-rowen** [nvaccess/nvda#11736]
I don’t know whether this issue should be reported to the NVDA Remote add-on or to NVDA. If I am wrong, please close this issue.
Thanks.

---

### Steps to reproduce:
1. Both parties install NVDARemote V2.3 add-ons.
2. The controlled party opens the NVDA advanced settings panel and selects "Attempt to cancel speech for expired focus events" as Yes.
3. The initiating control party and the controlled party discuss the host and key information to be used for connection.
4. After hearing the beep of successful connection, the initiating control party presses F11 to try to operate the other party's computer.

### Actual behavior:
NVDA does not announce anything.

### Expected behavior:
NVDA should correctly read the content on the controlled party's computer, just like operating this machine.

#### NVDA installed/portable/running from source:
installed

#### NVDA version:
NVDA2020.3RC1

#### Windows version:
Windows version: Windows 10 2004 (64-bit) build 19041.508

#### Name and version of other software in use when reproducing the issue:
NVDA Remote Version 2.3 

#### Other information about your system:
none

#### Does the issue still occur after restarting your computer?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
NVDA2020.2 also has this Issue.

#### If addons are disabled, is your problem still occuring?
Yes.

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
Yes.

**Comment by Adriani90** [comment]
cc: @feerrenrut, @tspivey, @josephsl

**Comment by feerrenrut** [comment]
I think this should be reported on the NVDA remote repo. Please note that this feature (Attempt to cancel speech for expired focus events) is currently a preview, but the goal is to turn it on by default in 2020.4

**Comment by feerrenrut** [comment]
Cross posting my comment from #11266:

Hopefully there is a simple fix we can apply on the NVDA side, otherwise we will defer the Cancellable speech by default change until the 2021.1 release. It should be noted that we don't believe we have broken any public add-on API with the Cancellable speech change, and if we do break the addon-API we expect add-on authors to help us to identify the breakage. Given the open and flexible nature of how NVDA add-ons interact with NVDA, we can only provide a best-effort approach with no strong guarantees. It is frustrating that NVDA Remote is written in a fragile way, relying on implementation details, and then maintenance falls on NV Access. The intention to make this option the default has been public for quite a while, and the option itself has existed for far longer. It seems NVDA Remote could be considered abandon-ware at this stage, it does not see active maintenance. Now would be a good time for someone new to take it on. The 2021.1 release will almost certainly break it.

**Comment by dingpengyu** [comment]
Is it possible for nvaccess to take over the maintenance of the NVDA remote plugin?
@feerrenrut

**Comment by lukaszgo1** [comment]
Looking at the error log it turns out that NVDA Remote fails when attempting to serialize FocusLossCancellableSpeechCommand using JSON. I don't think making `FocusLossCancellableSpeechCommand` serializable is easy as it also stores a reference to the NVDA object so options of solving this on the NVDA side seems to be pretty limited.
Log snipped for reference:

ERROR - unhandled exception (22:27:10.330) - MainThread (224):
Traceback (most recent call last):
  File "wx\core.pyc", line 3240, in <lambda>
  File "C:\Users\Admin\AppData\Roaming\nvda\addons\remote\globalPlugins\remoteClient\session.py", line 156, in speak
    priority=priority
  File "C:\Users\Admin\AppData\Roaming\nvda\addons\remote\globalPlugins\remoteClient\transport.py", line 123, in send
    obj = self.serializer.serialize(type=type, **kwargs)
  File "C:\Users\Admin\AppData\Roaming\nvda\addons\remote\globalPlugins\remoteClient\serializer.py", line 13, in serialize
    data = json.dumps(obj, cls=CustomEncoder).encode('UTF-8') + self.SEP
  File "json\__init__.pyc", line 238, in dumps
  File "json\encoder.pyc", line 199, in encode
  File "json\encoder.pyc", line 257, in iterencode
  File "C:\Users\Admin\AppData\Roaming\nvda\addons\remote\globalPlugins\remoteClient\serializer.py", line 31, in default
    return super().default(obj)
  File "json\encoder.pyc", line 179, in default
TypeError: Object of type FocusLossCancellableSpeechCommand is not JSON serializable

**Comment by feerrenrut** [comment]
Thanks for the investigation @lukaszgo1

NVDA remote should just ignore `FocusLossCancellableSpeechCommand` I believe. This is a design issue with NVDA remote. Attempting to intercept internal calls means it is tightly coupled to the implementation of those calls. 

Perhaps we could make it possible to be serialized and just get deserialized as a 'dummy' object.

**Comment by lukaszgo1** [comment]
> This is a design issue with NVDA remote. Attempting to intercept internal calls means it is tightly coupled to the implementation of those calls.

I agree that this is unfortunate but I don't think it can be written in any other way. Given how popular NVD Remote is I fell NV Access should really consider devoting some time to the native remote access solution in NVDA. For many people, especially in work environments NVDA is useless without one.

> Perhaps we could make it possible to be serialized and just get deserialized as a 'dummy' object.


Is it really  worth the  trouble? I mean not enabling cancellable speech by default for 2020.4 given how marginal improvements it brings seems to be small price to pay for not  breaking NVDA Remote.

**Comment by feerrenrut** [comment]
>  but I don't think it can be written in any other way

Another way would be to implement it as a speech and braille driver. 

> Is it really worth the trouble? I mean not enabling cancellable speech by default for 2020.4 given how marginal improvements it brings seems to be small price to pay for not breaking NVDA Remote.

This depends, for some cancellable speech is a big improvement (specifically in Gmail with Chrome). If we can fix it on the NVDA side with a serializable command, we won't break NVDA Remote. Otherwise we are going to break it in 2021.1 anyway.

**Comment by lukaszgo1** [comment]
> 
> 
> > but I don't think it can be written in any other way
> 
> Another way would be to implement it as a speech and braille driver.


In that case person whose computer is controlled would be unable to follow what controlling person is doing, as only one braille display driver/speech synthesizer can be active at the same time.

**Comment by cary-rowen** [comment]
I hope NVAccess will consider NVDA Remote as part of NVDA, which is really important.

**Comment by feerrenrut** [comment]
> In that case person whose computer is controlled would be unable to follow what controlling person is doing, as only one braille display driver/speech synthesizer can be active at the same time.

I was imagining that it would pass through to the configured speech / braille driver. It would act as a proxy for the real driver.

**Comment by seanbudd** [comment]
Closed via https://github.com/NVDARemote/NVDARemote/pull/237
